### PR TITLE
[#21132] refactor: improve send transaction content in wc flow

### LINKED ISSUE #21132: Improve the UX of the transactions modal
## User Story

[User Story](https://www.notion.so/dApp-interactions-ab1157cb960c4256b2c7650ce6c3d0d6?p=b67768dc9f5c455f9ffa6cec4c13a194&pm=s)

[User Story - Buy ETH Button](https://www.notion.so/dApp-interactions-ab1157cb960c4256b2c7650ce6c3d0d6?p=bb9c5b9e26964beb8925a7ecde9b6939&pm=c)

### Description

We should improve the UX of the transactions modal, where, as of now, we just show the raw json transaction (just formatting the values from hex to utf-8, so they’re human readable). 

[Design link](https://www.figma.com/design/mLyiYPXqu8KetdQONOlZpV/dApp-Interactions?node-id=1869-24859&t=NH2R1mGM4gZoF8UU-4)

## 

- [ ]  the user can see the structured data of the raw transaction
- [ ]  the user should see the dApp with its url and icon
- [ ]  the user should see the amount highlighted if not enough balance
- [ ]  the user should see the max time a transaction might take (estimated time)
- [ ]  the user should see the up-to-date “max time” and “max fees” while on the tx modal
    - the values should update every 10s

### PR DESCRIPTION
**Pull request by mohsen-ghafouri**
fixes #21132

### Summary

We should improve the UX of the transactions modal, where, as of now, we just show the raw json transaction (just formatting the values from hex to utf-8, so they’re human readable).

#### Areas that maybe impacted

- Send transaction in wallet connect

### Steps to test

1. Connect wallet to dapp
2. Try to sent ETH transaction

### Result

https://github.com/user-attachments/assets/fa9ca97b-0528-4096-801c-fe48b53fbf68

status:  ready

### CODE REVIEW

File: src/status_im/constants.cljs
Comment (clauxx) [inline src/status_im/constants.cljs]: better to use kebab-case here instead. If the value comes in snake_case, better to convert it I guess, just so we maintain the style consistency across the project.

File: src/status_im/contexts/wallet/common/utils.cljs
Comment (clauxx) [inline src/status_im/contexts/wallet/common/utils.cljs]: I guess if we get unknown, we should also show `>5 min` so it's not empty, but in the other cases I'd do it as in the designs, e.g:
`< 1` to `<1 min`
`< 3` to `1-3 min`
`< 5` to `3-5 min`

File: src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs
Comment (clauxx) [inline src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs]: I'd suggest to not reuse this event by passing the `refetch?` flag to avoid re-showing the modal (in general flags like this should be avoided in most cases), but instead to only re-use the `:effects.wallet-connect/prepare-transaction`. Since we also need the logic from `:wallet-connect/prepare-transaction-success` (but without the "show modal" dispatch), it can be extracted into a separate event and used for both preparing and updating the tx.

File: src/status_im/contexts/wallet/wallet_connect/utils/data_store.cljs
Comment (clauxx) [inline src/status_im/contexts/wallet/wallet_connect/utils/data_store.cljs]: can simplify this by using `(mapv identity data)`, which will achieve the same thing, also removing the need for  the util.

**Review by clauxx — CHANGES_REQUESTED** [review]
Thanks for the PR! Left some review comments. 

Other points to mention: 

1. Under the `Hex` tab, we should show the entire transaction hex-encoded (not the property values as hex).
2. I think would be best to hide the fields that are empty (have the `0x` value) and the internal ones the user doesn't need (`MultiTransactionId` and `Symbol`).
3. Also, I noticed this PR adds both the new transaction modal UI and the "Buy ETH" button. Would be good to include links to both user stories in the PR description (under testing notes maybe) so that the QA team can better understand the scope and how to test it.

File: src/status_im/constants.cljs
Comment (mohsen-ghafouri) [reply src/status_im/constants.cljs]: I used snake_case as In constants file i didn't see any kebab-case.  shouldn't we use the same style in constants file? @clauxx

File: src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs
Comment (mohsen-ghafouri) [reply src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs]: If i understand it correctly you mean to create a new event for update transaction like this? 

so `:wallet-connect/prepare-transaction-success` still get a flag to skip the modal part or not. right? 

```clojure
(rf/reg-event-fx
 :wallet-connect/update-transaction
 (fn [{:keys [db]}]
   (let [event    (data-store/get-db-current-request-event db)
         tx       (-> event data-store/get-request-params first)
         chain-id (-> event
                      (get-in [:params :chainId])
                      networks/eip155->chain-id)]
     (when tx
       {:fx [[:effects.wallet-connect/prepare-transaction
              {:tx         tx
               :chain-id   chain-id
               :on-success #(rf/dispatch [:wallet-connect/prepare-transaction-success % chain-id true])
               :on-error   #(rf/dispatch [:wallet-connect/on-processing-error %])}]]}))))
```

**Comment by mohsen-ghafouri** [comment]
> Thanks for the PR! Left some review comments.
> 
> Other points to mention:
> 
> 1. Under the `Hex` tab, we should show the entire transaction hex-encoded (not the property values as hex).
> 2. I think would be best to hide the fields that are empty (have the `0x` value) and the internal ones the user doesn't need (`MultiTransactionId` and `Symbol`).
> 3. Also, I noticed this PR adds both the new transaction modal UI and the "Buy ETH" button. Would be good to include links to both user stories in the PR description (under testing notes maybe) so that the QA team can better understand the scope and how to test it.

for point 1, this is exactly what we get from dapp, did i miss something? the only operation we do on it is to stringify and display it. @clauxx

File: src/status_im/constants.cljs
Comment (clauxx) [reply src/status_im/constants.cljs]: I think we use snake case for strings that come from external sources e.g JSON-RPCs like `eth_sendTransaction` and we hide them behind a constant, so the project is still consistent with the naming convention. Here we have keywords with underscores though, so it could be that we will use e.g. ` :less_than_five_minutes` outside the constants file, which will not be consistent. Since we already do the conversion from string to keyword somewhere, might as well convert it to kebab case as well.

File: src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs
Comment (clauxx) [reply src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs]: So this is one way, but another way would be to have separate events for the success handlers, but still share the logic, either with a common event for updating the state but without showing the modal, or just a utility function. The the success event can be passed to the `prepare-transaction` event, or alternatively have different events for preparing and updating the transaction.

This is just off the top of my head, so I might be missing things, but the point is that there are ways to avoid passing the flag, even if we end up with a few more events. Passing the flag through two events just to avoid showing the modal will lead to other issues in the future. If you're stuck, we can go through it in a call at some point and figure it out together.

**Comment by clauxx** [comment]
> > Thanks for the PR! Left some review comments.
> > Other points to mention:
> > 
> > 1. Under the `Hex` tab, we should show the entire transaction hex-encoded (not the property values as hex).
> > 2. I think would be best to hide the fields that are empty (have the `0x` value) and the internal ones the user doesn't need (`MultiTransactionId` and `Symbol`).
> > 3. Also, I noticed this PR adds both the new transaction modal UI and the "Buy ETH" button. Would be good to include links to both user stories in the PR description (under testing notes maybe) so that the QA team can better understand the scope and how to test it.
> 
> for point 1, this is exactly what we get from dapp, did i miss something? the only operation we do on it is to stringify and display it. @clauxx

The `:wallet_buildTransaction` RPC to status-go will also return the `RawTx` property, which is the transaction encoded as hex.

File: src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs
Comment (mohsen-ghafouri) [reply src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs]: Honestly, I don’t see any issue with extending a function by adding an option to reuse it, as long as it doesn’t cause breaking changes. This approach is better than duplicating logic under different event names. I’ve noticed similar cases where the function content is identical, but only the event name differs, which isn’t ideal. Extending a function still aligns with the Open/Closed Principle and allows us to reuse logic without duplication. 

wdyt? @shivekkhurana @ilmotta

**Comment by mohsen-ghafouri** [comment]
`:wallet_buildRawTransaction`, is that content useful for user? as there wasn't any example of it in design i couldn't understand the purpose of it, maybe @xAlisher can share your thoughts here?

**Comment by clauxx** [comment]
> > Thanks for the PR! Left some review comments.
> > Other points to mention:
> > 
> > 1. Under the `Hex` tab, we should show the entire transaction hex-encoded (not the property values as hex).
> > 2. I think would be best to hide the fields that are empty (have the `0x` value) and the internal ones the user doesn't need (`MultiTransactionId` and `Symbol`).
> > 3. Also, I noticed this PR adds both the new transaction modal UI and the "Buy ETH" button. Would be good to include links to both user stories in the PR description (under testing notes maybe) so that the QA team can better understand the scope and how to test it.
> 
> for point 1, this is exactly what we get from dapp, did i miss something? the only operation we do on it is to stringify and display it. @clauxx

That's a good point, we can only get the raw hex transaction once we have the signature, which we get after the user authenticates it by sliding. We could just hex encode the args JSON string, but I wonder if that has value to the user. @shivekkhurana @xAlisher what is the value from the hex tab and what is the use case for it?

File: src/status_im/contexts/wallet/wallet_connect/utils/data_store.cljs
Comment (mohsen-ghafouri) [reply src/status_im/contexts/wallet/wallet_connect/utils/data_store.cljs]: `mapv` returns a vector, whereas `map` returns a lazy sequence, and here to pass to flat-list i need lazy sequence

File: src/status_im/constants.cljs
Comment (mohsen-ghafouri) [reply src/status_im/constants.cljs]: Just to clarify, the estimation values are received on the mobile side as integers. I created this constant to match the same enum and also to serve as documentation. Therefore, using kebab-case is not necessary in this case IMO.

Comment (clauxx) [reply src/status_im/constants.cljs]: > I used snake_case as In constants file i didn't see any kebab-case

> Therefore, using kebab-case is not necessary in this case IMO.

I'm confused. AFAIK kebab-case is the convention for clojure and I couldn't find any examples of snake_case

File: src/status_im/contexts/wallet/wallet_connect/utils/data_store.cljs
Comment (clauxx) [reply src/status_im/contexts/wallet/wallet_connect/utils/data_store.cljs]: you're right, then just `(map identity data)` will do

File: src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs
Comment (clauxx) [reply src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs]: I never said we should duplicate logic though, I just said we should avoid using a flag as an argument to handle the additional "update" scenario.

File: src/status_im/constants.cljs
Comment (mohsen-ghafouri) [reply src/status_im/constants.cljs]: Sorry if my explanation is not clear enough, the point is we could directly use input from rpc in the format function , but i though number alone can confuse us in future, so i just created the same enum as go side has. 
everything in constants file is snake_case. am i missed something? 
without constant it looks like this: 
```clojure
(defn estimated-time-format
  "Formats the estimated time for a transaction"
  [estimated-time]
  (condp = estimated-time
    1              "> 5"
    2              "< 1"
    3              "1-3"
    4              "3-5"
    "> 5"))
```

File: src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs
Comment (mohsen-ghafouri) [reply src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs]: I'm confused, if can't pass an argument to a func to extend it, how we can avoid duplication, maybe if you can give me some example that would be nice.

File: src/status_im/constants.cljs
Comment (clauxx) [reply src/status_im/constants.cljs]: I understand that what we use is the number and agree that using keywords makes it more understandable. My point is only about changing e.g. `:less_than_one_minute` (snake case) to `:less-than-one-minute` (kebab case)

> everything in constants file is snake_case 

We have no variable names or keywords that use snake case in the constants namespace or the project (from what I saw), and if there are then it's likely there by mistake

File: src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs
Comment (ilmotta) [reply src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs]: I don't have any strong preference guys, either way can work well and has pros/cons.

Side note? the solution in `:wallet-connect/prepare-transaction-success` to dispatch `:wallet-connect/show-request-modal` is more of a UI concern. This is not a problem from this PR or in general recent PRs, but a thing we tend to do quite often in the code base, more than we should perhaps. This could be one of the reasons why it's difficult to model some of our events, because they are caring about UI concerns that the view should decide.

File: src/status_im/constants.cljs
Comment (mohsen-ghafouri) [reply src/status_im/constants.cljs]: Oh my bad, sorry i so confused with kebab case 😓, yes that's right it should be kebab-case, i will fix it.

File: src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs
Comment (mohsen-ghafouri) [reply src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs]: yeah the issue it this event will be called in the background once it receive an event from dapp, and i can't think any other way to achieve the same result without duplicating the logic to avoid using flag.

Comment (clauxx) [reply src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs]: we could extract the logic from the events into an external function that would be used by the events to avoid duplicating the logic, but since there's no strong preferences, I won't push on this

File: src/status_im/contexts/wallet/common/utils.cljs
Comment (clauxx) [inline src/status_im/contexts/wallet/common/utils.cljs]: maybe without the spaces here, as `> 5 min` looks a bit weird. Usually the "smaller than" is shown right next to the number like `>5 min`

File: src/status_im/contexts/wallet/wallet_connect/utils/transactions.cljs
Comment (clauxx) [inline src/status_im/contexts/wallet/wallet_connect/utils/transactions.cljs]: remove log

**Review by clauxx — APPROVED** [review]

File: src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs
Comment (mohsen-ghafouri) [reply src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs]: Without an example, it’s hard to fully grasp the implementation you’re suggesting, but I’ll try to see if I can use the logic without adding flags or causing duplication.

Comment (mohsen-ghafouri) [reply src/status_im/contexts/wallet/wallet_connect/events/session_requests.cljs]: Thanks for your feedback @clauxx, I applied some changes, could you please recheck?

**Comment by mariia-skrypnyk** [comment]
Hey @mohsen-ghafouri !

Can this PR be tested or you are waiting for more reviews?

**Comment by mohsen-ghafouri** [comment]
Hi @mariia-skrypnyk yes let's move forward with this PR, if i get any feedback from Alisher, i will apply in next PR. so you can test this PR. thanks

**Comment by mariia-skrypnyk** [comment]
Hey @mohsen-ghafouri !

Have several questions regarding scope of your PR and logic of Acceptance criteria's.

1) Is it correct that I should test buying assets if we have not enough amount of them for transaction and for fee?
2) If the 1) is "yes" then how can I trigger such transaction for example on uniswap if in case not enough assets uniswap doesn't allow me to do it:

Screenshot 2024-09-27 at 10 27 14

3) in what format do we have data that we receive? it doesn't look informative from user perspective I think as I am waiting the same data as I see on uniswap modal here:

Screenshot 2024-09-27 at 10 31 18
Screenshot 2024-09-27 at 10 33 39

Thanks!

**Comment by mariia-skrypnyk** [comment]
During my PR testing I've caught 3-4 crashes.
They were on PR build that I installed firstly and when I re-installed it I also caught 1-2 of them.
All crashes found me on sign transactions WC modals when I was doing some amount changes on the dAap side.
Unfortunatelly crashed are not 10/10 times reproducible but I attached logs (hope it will help).

Steps:
1. Login to app
2. Change wallet settings to the Mainnet
3. Go to Wallet -> Account
4. Connect Account to Uniswap dApp
5. Create any swap transaction
6. While the transaction modal appeared try to change some amounts and tokens on Uniswap site.
7. Status crashes in several minutes.

[crash_WC_logs.zip](https://github.com/user-attachments/files/17161721/crash_WC_logs.zip)

**Comment by mohsen-ghafouri** [comment]
Hello @mariia-skrypnyk, scope of this PR is just about display "Buy ETH" button if balance is less than transaction amount/fee. and nothing to do with buy/swap flow. 
for point 3 this is how we agreed on design, we should well formatted in "Max fees"

**Comment by mariia-skrypnyk** [comment]
@mohsen-ghafouri thanks

**Comment by mohsen-ghafouri** [comment]
I couldn't reproduce the issue, user doesn't have ability to change content of sent transaction, if you close modal and change amount in dapp, then it will be consider as new request. so user first have to do action about previous one then new request will appear.

**Comment by mariia-skrypnyk** [comment]
Thanks @mohsen-ghafouri .

Will check one more time and run tests.

**Comment by mariia-skrypnyk** [comment]
Hi @mohsen-ghafouri !

Builds were rechecked, saw no crash this time.
Failed e2e is not related.

Please merge your PR.

**Comment by mohsen-ghafouri** [comment]
I Will create a follow up issue to improve HEX tab (to include raw hex data)

**Pull request merged**
